Document Visual Studio (#999)
longturn/freeciv21: docs/Contributing/visual-studio.rst

Setting up Visual Studio
************************

Freeciv21 can be compiled using Microsoft Visual Studio |reg| and :file:`clang`. This page will help you
get version 2022 up and running.

.. warning:: Visual Studio and the corresponding dependencies require a great deal of HDD space on your
    computer. Be sure to have at least 40GB of available space before starting this process.

Base Installation
=================

Start by `downloading <https://visualstudio.microsoft.com/vs/community/>`_ the Community Edition installer.

Double-click the :file:`VisualStudioSetup.exe` file in your Downloads directory and follow the prompts to
get the setup process started. At some point you will be prompted to select the type of workload you want to
install. Select :guilabel:`Desktop development with C++` and :guilabel:`Python Development`.  Next click on the
Individual components tab and select the following options: :guilabel:`Git for Windows`,
:guilabel:`C++ Clang Compiler for Windows (13.0.0)`, :guilabel:`C++ Clang-cl for v143 build tools (x64/x86)`,
:guilabel:`Python 3 64-bit (3.9.7)`, and :guilabel:`Python native development tools`.

> When ready click :guilabel:`Install`. Depending on your Internet connection speed, this may take awhile.

Reviewer: ```suggestion
When ready click :guilabel:`Install`. Depending on your Internet connection speed, this may take a while.
```

> When the installation process is finished you will be presented with a collection of options. Goto the very

Reviewer: Pun intended?
```suggestion
When the installation process is finished you will be presented with a collection of options. Go to the very
```

bottom right and click the link to :strong:`continue without code`. You can now close the Visual Studio
installer. Leave the main Visual Studio IDE open, but you can minimize it for the next step.

> Setup VCPKG
> ===========

Reviewer: ```suggestion
Set up VCPKG
============
```
Or `VCPKG setup`

:file:`vcpkg` is a library package manager developed by Microsoft. It makes downloading and installing
third-party libraries much easier. Freeciv21 uses :file:`vcpkg` in this way via the manifest file
:file:`vcpkg.json`. Documentation is available here: https://github.com/microsoft/vcpkg

First, create a directory on your computer to install :file:`vcpkg` into. For example: :file:`C:\\Tools`.
The directory can be anywhere, however the author prefers to :strong:`not` install things in the root of the
:file:`C:\\` drive.

> Open an Administrative elevated PowerShell terminal window. Easiest way to do this is to right-click on the

Reviewer: ```suggestion
Open an Administrative elevated PowerShell terminal window. The easiest way to do this is to right-click on the
```
Not sure about this one

> Start Menu and select it from the menu: ``Windows PowerShell (Admin)``

Reviewer: `:guilabel:`? (for Start and PowerShell)

.. code-block:: rst

    Tools> git clone https://github.com/microsoft/vcpkg
    Tools> .\vcpkg\bootstrap-vcpkg.bat
    Tools> setx VCPKG_ROOT C:/Tools/vcpkg /M
    Tools> exit

GitHub
======

The Longturn.net Community uses the online source code control and revision system known as
`GitHub <https://github.com/>`_. To contribute, you will need an account on this platform. There is no cost.

With an account, you can go to the `Longturn <https://github.com/longturn>`_ community repository page and
:strong:`fork` a repository (such as the Freeciv21 repository) to your personal GitHub account. Go to the main
page of the repository you want to fork and you will find a :strong:`fork` button in the upper-right corner.

In order to get code pushed to the forked repository to your local workstation, you need to setup an
SSH key pair to share with GitHub. Follow these
`instructions <https://docs.github.com/en/authentication/connecting-to-github-with-ssh>`_.

With that set up, now it's time to clone the forked repository from your personal GitHub account to a local
copy on your workstation. The typical way to do this is with the :code:`https` protocol. However, this only

> works if you want to download a copy of a repository and not push any changes back up. To do that, you have to

Reviewer: > and not push any changes back up.

Maybe this is too much git jargon? "publish/contribute your changes back"?

use the :code:`ssh` protocol instead.

Now you want to clone the respository. You can get the appropriate command by going to your forked copy in a
browser, click the code button and then select the SSH option as shown in this sample screenshot:

.. image:: ../_static/images/github_clone_ssh.png
    :align: center
    :height: 250
    :alt: GitHub Clone SSH

> Bring up Visual Studio. Select :menuselection:`View --> Terminal` to open a terminal in the IDE (if one is not

Reviewer: ```suggestion
Bring up Visual Studio. Select :menuselection:`View --> Terminal` to open a terminal in the IDE if one is not
```

already shown at the bottom. Once you have the proper GitHub path, here is the command to clone the repository:

.. code-block:: rst

    PS C:\Users\[user]\Source\Repos> git clone git@github.com:[username]/freeciv21.git

This will clone the forked repository to the :file:`C:\\Users\\[User]\\Source\\Repos\\freeciv21` directory on
your computer.

The final repository setup item is to link the original Longturn project repository to your local area on
your computer:

.. code-block:: rst

    Repos> cd freeciv21
    Repos\freeciv21> git remote add upstream https://github.com/longturn/freeciv21.git
    Repos\freeciv21> git fetch upstream
    Repos\freeciv21> git fetch upstream master

You will also need to set a couple global configuration settings so :code:`git` knows a bit more about you.

.. code-block:: rst

    freeciv21> git config --global user.email "[email address associated with GitHub]"
    freeciv21> git config --global user.name "[your first and last name]"

Build Visual Studio Solution
============================

> Now let's get Visual Studio setup. Select :menuselection:`Git --> Local Repositories --> Open Local Repository`

Reviewer: ```suggestion
Now let's get Visual Studio set up. Select :menuselection:`Git --> Local Repositories --> Open Local Repository`
```
(past participle)

and then select the :file:`freeciv21` folder. Visual Studio will take a minute to parse the source tree.

Select :menuselection:`File --> Open --> Folder`. Select the :file:`freeciv21` directory and then Visual Studio
will parse the files in there. It can take a few mintues for this to complete. When complete, Visual Studio
will open a :strong:`CMake Overview Pages` tab.

Microsoft provides documentation on CMake in Visual Studio --
https://docs.microsoft.com/en-us/cpp/build/cmake-projects-in-visual-studio?view=msvc-170

Final Steps and Notes
=====================

As of this writing, there is not a fully integrated setup for configuring, compiling (building) and debugging
Freeciv21 inside of the Visual Studio user interface. If you open any file (such as :file:`CMakeLists.txt`)
and make a change, Visual Studio will prompt to :strong:`Generate` updated C++ Intellisense information. If
selected, Visual Studio will configure using the ``fullrelease`` preset in :file:`CMakePresets.json` using the
Ninja generator. Ninja only supports MSVC in Visual Studio. However, the Longturn community does not support

> MSVC due to licensing constraints. Instead the Longturn community supports the open source LLVM Clang-Cl

Reviewer: This makes it sound like ClangCL is fine from a licensing point of view, but I think it uses the Microsoft C++ runtime by default. The real reason for using Clang is that our code require proprietary extensions that MSVC doesn't implement.
